### PR TITLE
External database configuration

### PR DESCRIPTION
**Pull request by kfox1111**
This patch makes all the database settings configurable.

fixes: https://github.com/spiffe/helm-charts/issues/37

### CODE REVIEW

File: charts/spire/charts/spire-server/values.yaml
Comment (faisal-memon) [inline charts/spire/charts/spire-server/values.yaml:233]: I think this should probably go under the persistence of datastore sections

Comment (kfox1111) [reply charts/spire/charts/spire-server/values.yaml:233]: Thanks for the review.

Unfortunately, a nested chart can't contain any nestings in the values file of the chart its nested into.
so can be anything at the root of the chart, such as:
```
postgresql:
  enabled: true
```

but cant be like:
```yaml
dataStore:
  sql:
     postgresql:
       enabled: true
```

Its a limitation of helm itself.

**Comment by kfox1111** [comment]
Split out the internal db support to its own pr: https://github.com/spiffe/helm-charts/pull/233

Should be ready for review.

File: charts/spire/charts/spire-server/templates/secret.yaml
Comment (marcofranssen) [inline charts/spire/charts/spire-server/templates/secret.yaml:103]: Why are we switching this to a secret?

Can't we just put the secrets in a secret.

E.g.:

We could store the db creds in a secret and load them into the respective env var, this way we don't need to change the configmap:

See
http://www.postgresqlnow.com/connect-env-variables/

File: charts/spire/charts/spire-server/values.yaml
Comment (marcofranssen) [inline charts/spire/charts/spire-server/values.yaml:90]: For documentation purposes and less hidden magic it might be better to default this to sqlite.

```suggestion
dataStore:
  sql:
    # -- Set any setting from https://github.com/spiffe/spire/blob/v1.6.3/doc/plugin_server_datastore_sql.md under this section.
    plugin_data:
      # -- Other supported databases are "postgres" and "mysql"
      database_type: sqlite3
      # -- The connectionstring to connect to your database instance.
      connection_string: /run/spire/data/datastore.sqlite3
```

File: examples/external-mysql/values.yaml
Comment (marcofranssen) [inline examples/external-mysql/values.yaml:6]:
```suggestion
        host: mysql
        port: 3306
        database: spire
        username: spire
        password: CHANGEME
        options:
          - parseTime: true
```

Too support the confimap combination with secrets we might want to more stronglytype the values file.

File: examples/external-postgresql/values.yaml
Comment (marcofranssen) [inline examples/external-postgresql/values.yaml:6]:
```suggestion
        host: postgresql
        port: 5432
        database: spire
        username: spire
        password: CHANGEME
        options: 
          - sslmode: disable
```

Too support the confimap combination with secrets we might want to more strongly type the values file.

**Review by marcofranssen — CHANGES_REQUESTED** [review]
Couple of suggestions to follow patterns used often in other charts I have seen in the past.

File: charts/spire/charts/spire-server/values.yaml
Comment (kfox1111) [reply charts/spire/charts/spire-server/values.yaml:90]: The logic here was developed with the idea that internal databases as per https://github.com/spiffe/helm-charts/pull/233 could be a thing. The suggested settings somewhat conflict with that.

File: examples/external-mysql/values.yaml
Comment (kfox1111) [reply examples/external-mysql/values.yaml:6]: What happens when they add a new database type that isn't sqllite or a sql server? say, mongo or some other nosql database? Are we backing ourselves into a corner trying to interpret connection_string?

If we're adding config options I don't think they should be under plugin_data, which should match closely what is in spire's config. We probably should move it up a level to under sql: if we do that.

File: charts/spire/charts/spire-server/templates/secret.yaml
Comment (marcofranssen) [reply charts/spire/charts/spire-server/templates/secret.yaml:103]: To some extend it does protect yes, in the end k8s secrets are just obfuscated as base64.

In many cases we pull secrets from external sources injected as env variables, having the entire config as secret complicates this strategy.

File: examples/external-mysql/values.yaml
Comment (marcofranssen) [reply examples/external-mysql/values.yaml:6]: In that case we probably have to add another implementation that builds the connection string based on that database type

File: charts/spire/charts/spire-server/templates/secret.yaml
Comment (kfox1111) [reply charts/spire/charts/spire-server/templates/secret.yaml:103]: Users of k8s often confuse encoding and encryption or assume because one is used it means its designed to be obscured in favor over encryption. This is **not** the case in Kubernetes. base64 was chosen to encode secret data as its often binary data and needs to be put into a yaml/json document which isn't binary safe. But encryption is supported too. In many production systems, secrets are encrypted via:
https://kubernetes.io/docs/tasks/administer-cluster/encrypt-data/

The NSA's recommendation for Kubernetes hardening mentions it explicitly. I believe the DOD and CIS benchmark recommendations are also to turn on encryption for secrets.

My own clusters are encrypted. Secrets are often encrypted while configmaps are often not.

---

As for arbitrary plugin support, that's a debate on governance. I'm on the side of, if the spire folks needed plugin support to support the users, I'm for supporting the users in their needs as well. We'll hash that problem out in the governance meeting.

File: examples/external-mysql/values.yaml
Comment (kfox1111) [reply examples/external-mysql/values.yaml:6]: I guess in the case of internal databases, we can just use enable to flag if we totally ignore the config there. I'll take a stab at it.

File: .github/tests/charts.json
Comment (marcofranssen) [inline .github/tests/charts.json:18]:
```suggestion
```

File: .github/tests/charts.json
Comment (marcofranssen) [inline .github/tests/charts.json:24]:
```suggestion
```

File: .github/tests/production-external-mysql/install.sh
Comment (marcofranssen) [inline .github/tests/production-external-mysql/install.sh:22]:
```suggestion
cat <<EOF > /tmp/$$-spire-values.yaml
```

File: examples/external-postgresql/values.yaml
Comment (marcofranssen) [inline examples/external-postgresql/values.yaml:12]:
```suggestion
```

File: charts/spire/charts/spire-server/README.md
Comment (faisal-memon) [inline charts/spire/charts/spire-server/README.md:90]: What do you think of keep this together with the rest of the datastore config?

File: charts/spire/charts/spire-server/values.yaml
Comment (faisal-memon) [inline charts/spire/charts/spire-server/values.yaml:93]: Maybe comment ofer this section that this config is only needed for postgres or mysql.

File: charts/spire/charts/spire-server/values.yaml
Comment (faisal-memon) [inline charts/spire/charts/spire-server/values.yaml:92]: Should the password be read from the secret?

File: charts/spire/charts/spire-server/templates/secret.yaml
Comment (faisal-memon) [reply charts/spire/charts/spire-server/templates/secret.yaml:103]: Prefer to only have the password be a k8s secret

Comment (kfox1111) [reply charts/spire/charts/spire-server/templates/secret.yaml:103]: Ok. I'll prepare another pr with that in it so we can review.

Struggling a bit with how all this reconciles with the principal of keeping things "simple". All this combined is the most complicated way of setting "connection_string" possible.

File: charts/spire/charts/spire-server/README.md
Comment (kfox1111) [reply charts/spire/charts/spire-server/README.md:90]: It doesn't have a 1:1 relationship with dataStore. 

Its potentially used by sqlite3 for spire-server, by the disk KeyStore, and/or by Tornjak.

So when mysql/postgresql is used that has its own storage, it still may be needed.

File: charts/spire/charts/spire-server/templates/secret.yaml
Comment (kfox1111) [reply charts/spire/charts/spire-server/templates/secret.yaml:103]: Here it is. https://github.com/spiffe/helm-charts/pull/255

File: charts/spire/charts/spire-server/values.yaml
Comment (kfox1111) [reply charts/spire/charts/spire-server/values.yaml:92]: Are you asking for the option of reading the password out of an existing k8s secret?

File: charts/spire/charts/spire-server/README.md
Comment (kfox1111) [reply charts/spire/charts/spire-server/README.md]: See https://github.com/spiffe/helm-charts/pull/211
As its not merged, and this kind of depends on it, some of the changes appear here too.

File: charts/spire/charts/spire-server/values.yaml
Comment (kfox1111) [reply charts/spire/charts/spire-server/values.yaml]: Done

File: charts/spire/charts/spire-server/README.md
Comment (faisal-memon) [inline charts/spire/charts/spire-server/README.md:87]: Why did this get all condensed?

**Comment by kfox1111** [comment]
Somehow that merge still made a dco issue. :/ Going to manually squash...

File: charts/spire/charts/spire-server/README.md
Comment (kfox1111) [reply charts/spire/charts/spire-server/README.md]: Merge issue. Should be fixed.

File: examples/external-mysql/values.yaml
Comment (marcofranssen) [inline examples/external-mysql/values.yaml:9]: Maybe add a disclaimer, folks probably want to inject this differently so it doesn't end up in git repo. To also mitigate @faisal-memon his question.

**Review by marcofranssen — CHANGES_REQUESTED** [review]
Few small nits.

File: charts/spire/charts/spire-server/README.md
Comment (marcofranssen) [reply charts/spire/charts/spire-server/README.md]: This gets condensed not because of merge issue, but because of the helm docs introduced.

See https://github.com/norwoodj/helm-docs#valuesyaml-metadata for more details.

In essence adding a documentation tag on any of the nested properties would resolve this again.

**Comment by marcofranssen** [comment]
> Somehow that merge still made a dco issue. :/ Going to manually squash...

Thats because merges don't automatically add the signof. It is a missing feature in github for merging prs into prs. You could have added the dco during the merge dialog by typing it yourself. However, the squash is also fine in this case 👍

File: charts/spire/charts/spire-server/README.md
Comment (kfox1111) [reply charts/spire/charts/spire-server/README.md]: No, there is that other persistence pr, and I manually merged part of it here to get them looking the same again, but put it in the wrong place. :/

I fixed the manual merge issue location and it should be back to split out again.

File: examples/external-mysql/values.yaml
Comment (kfox1111) [reply examples/external-mysql/values.yaml:9]: oh, great point... It should be removed here and documentation updated.... Will fix.

Comment (kfox1111) [reply examples/external-mysql/values.yaml]: Fix committed.

**Comment by kfox1111** [comment]
> Thats because merges don't automatically add the signof. It is a missing feature in github for merging prs into prs. You could have added the dco during the merge dialog by typing it yourself. However, the squash is also fine in this case +1

I could have sworn I reviewed that and verified my signoff was in it. but somehow it still failed. But, yeah. fixed for now.

File: charts/spire/charts/spire-server/README.md
Comment (kfox1111) [reply charts/spire/charts/spire-server/README.md:88]: And a few advantages. less volumes for admins to have to snapshot/backup.

Comment (faisal-memon) [reply charts/spire/charts/spire-server/README.md:88]: I agree with @kfox1111 on having 1 volume. Keeps things simple. If we need to have multiple volumes we can add in the future.

File: charts/spire/charts/spire-server/templates/_helpers.tpl
Comment (kfox1111) [reply charts/spire/charts/spire-server/templates/_helpers.tpl:108]: So, 
{{- define "spire-server.config-mysql-options" }}
?

The general idea I was thinking, was to have the reader of the code pay attention a little bit to the difference between the two. the function is converting the options into a url query in the one case, but a key=value (more straight through options case?) when it comes to postgresql.

Another option might be to spell that out more explicitly...
"spire-server.config-mysql-options-2-url-query"
"spire-server.config-postgresql-options-2-key-value" ?

File: charts/spire/charts/spire-server/templates/_helpers.tpl
Comment (kfox1111) [reply charts/spire/charts/spire-server/templates/_helpers.tpl:113]: If it was just the = it would be a lot easier. the = and the urlquery makes it much trickier. Not sure how to reduce it without being very similar in line count (maybe slightly less) but with more convolution. The printf join line is a bit different too.

Comment (faisal-memon) [reply charts/spire/charts/spire-server/templates/_helpers.tpl:113]: got it, no need to change anything. thanks for the explanation.

File: charts/spire/charts/spire-server/README.md
Comment (faisal-memon) [reply charts/spire/charts/spire-server/README.md:88]: Can we auto set the `persistence` if `sqllite3` is the db or `keymanager` is `disk`?

File: charts/spire/charts/spire-server/values.yaml
Comment (faisal-memon) [inline charts/spire/charts/spire-server/values.yaml:86]:
```suggestion
    databaseType: sqlite3
```

File: charts/spire/charts/spire-server/values.yaml
Comment (faisal-memon) [inline charts/spire/charts/spire-server/values.yaml:87]:
```suggestion
    databaseName: spire
```

File: charts/spire/charts/spire-server/templates/_helpers.tpl
Comment (faisal-memon) [inline charts/spire/charts/spire-server/templates/_helpers.tpl:148]: Will the rendered configMap have the password visible in it?

Comment (faisal-memon) [reply charts/spire/charts/spire-server/templates/_helpers.tpl:148]: nm i see it now, it pulls from the env var.

File: charts/spire/charts/spire-server/values.yaml
Comment (faisal-memon) [inline charts/spire/charts/spire-server/values.yaml:89]:
```suggestion
    # -- Likely 5432 for postgres and 3306 for mysql
```

File: charts/spire/charts/spire-server/values.yaml
Comment (faisal-memon) [inline charts/spire/charts/spire-server/values.yaml:86]:
```suggestion
    database_type: sqlite3
    
    # The remaining config in this section only applies to postgres and mysql databases

```

File: charts/spire/charts/spire-server/values.yaml
Comment (faisal-memon) [inline charts/spire/charts/spire-server/values.yaml:94]:
```suggestion
    options: []
```

File: charts/spire/charts/spire-server/values.yaml
Comment (faisal-memon) [inline charts/spire/charts/spire-server/values.yaml:96]:
```suggestion
    # -- Set any setting from https://github.com/spiffe/spire/blob/main/doc/plugin_server_datastore_sql.md under this section.
```

Comment (faisal-memon) [reply charts/spire/charts/spire-server/values.yaml:96]: So we don't have to keep the version up to date.

File: charts/spire/charts/spire-server/values.yaml
Comment (kfox1111) [reply charts/spire/charts/spire-server/values.yaml:94]: I can't find the comment now. probably lost in the squash, but I believe the comment there was requested by Marco.

Comment (kfox1111) [reply charts/spire/charts/spire-server/values.yaml:97]: oh... I see what you did... this with the other patch puts a comment on the whole section instead...

But that isn't quite equivalent, because, while you can see the comment when looking a the raw values, it doesn't show up in the README. Not sure how to do the same thing in a way that would show up in the README

**Comment by kfox1111** [comment]
Added port auto setting, so if the user is using the default db ports, it will "just work"

File: charts/spire/charts/spire-server/values.yaml
Comment (faisal-memon) [reply charts/spire/charts/spire-server/values.yaml:97]: Yea, not sure if thats possible with helm docs. Can alternatively add the only used by to each of the variable if you want it to show up in the readme.

File: charts/spire/charts/spire-server/README.md
Comment (faisal-memon) [reply charts/spire/charts/spire-server/README.md:88]: Whats the value of disabling persistence?

File: charts/spire/charts/spire-server/templates/_helpers.tpl
Comment (faisal-memon) [reply charts/spire/charts/spire-server/templates/_helpers.tpl:108]: Its fine with me to keep as is. seems like its doing something different.

File: charts/spire/charts/spire-server/README.md
Comment (kfox1111) [reply charts/spire/charts/spire-server/README.md:88]: I've used it in several, related ways on other charts (there may be others)

1. In ci pipelines to speed up deployment of testing your own software, that uses spire. Provisioning the volume takes a bit to do, and being able to skip it reduces the overall runtime of the pipeline.  This enabled more rapid development/testing cycles.
2. In a ci pipeline where your deploying a test k8s. Similar to 1, not needing to provision a persistent volume driver at all can speed up deployment of the test cluster. Most valuable when dynamically allocating clusters in the pipeline
3. Some ci systems don't clean up after themselves very well (EX: cancellations). Its much easier to clean up helm charts that don't have pvc's.
4. Complex testing: while minikube has pv support in standalone node mode, it does not in its multinode mode. So if you want to do some nice multinode testing with it, just disabling persistence lets you deploy test instances of things such as spire, mysql, your code, etc in a nice scailable way, with stuff like antiaffinity, for quick testing of something that looks production like.

File: examples/bin/readpw.sh
Comment (marcofranssen) [inline examples/bin/readpw.sh:1]:
```suggestion
#!/usr/bin/env bash

# shellcheck shell=bash disable=SC2034
```

**Review by marcofranssen — COMMENTED** [review]
Few more nitpicks, disclaimer, reviewed from the mobile app, hopefully suggestions without typos.

File: examples/bin/readpw.sh
Comment (kfox1111) [reply examples/bin/readpw.sh:1]: I did not put a shell on it because its not really executable. It can only be sourced. If you execute it, it runs, but the env var gets undefined at the end.

File: charts/spire/charts/spire-server/README.md
Comment (marcofranssen) [reply charts/spire/charts/spire-server/README.md:88]: Due the way these PVCs work any other workloads then spire-server will already define their own `persistentVolumeClaim`

See spire-server statefulset.

```yaml
  volumeClaimTemplates:
    {{- if eq (.Values.persistence.enabled | toString) "true" }}
    - metadata:
        name: spire-data
      spec:
        accessModes:
          - {{ .Values.persistence.accessMode | default "ReadWriteOnce" }}
        resources:
          requests:
            storage: {{ .Values.persistence.size }}
        {{- if .Values.persistence.storageClass }}
        storageClassName: {{ .Values.persistence.storageClass }}
        {{- end }}
    {{ end }}
```

If tornjak needs a PVC it will have to implement its own meaning you would also want to be able to choose its own storageClass and accessModes etc. That is just how it works. Going with this API simply means we will have to break the API as soon we add a PVC to tornjak. IMHO the persistence flag is to generalized, hence why it previously was at dataStorage specifically for the dataStorage volume.

maybe it should be 

```
dataStore:
  persistence:
    enabled: true
    …
```

So we can later do things like

```
tornjak:
  persistence:
    enabled: true
```

Oh and if we are about to disable persistence, then probably the statefulset should also be changed to a deployment if persistence is disabled. That is something I have seen in other charts as well. Happy to create a PR to patch those changes on top of this PR.

Comment (kfox1111) [reply charts/spire/charts/spire-server/README.md:88]: > maybe it should be
> ```yaml
> dataStore:
>  persistence:
>    enabled: true
>    …
> ```
> So we can later do things like
> 
> ```yaml
> tornjak:
>  persistence:
>    enabled: true
> ```
and one for the disk keystore too. But, yeah, there'd be 3x the storage waste I think. I don't see the benefit in having so many volumes, especially once you have multiple replicas. But there is benefit in combining them for the end user.

> Oh and if we are about to disable persistence, then probably the statefulset should also be changed to a deployment if persistence is disabled. That is something I have seen in other charts as well. Happy to create a PR to patch those changes on top of this PR.

Other then for testing purposes, I don't think the user should ever be disabling persistence on the disk keystore. So at least 1 pvc is needed. Having the test case deployment look like the prod deployment in every way but deploy a little faster and not save state permanently is the ideal, so it to still be a statefulset is useful. Its totally ok for it to remain a statefulset and have no volumes. There are other properties of a statefulset that are useful outside pvs, such as persistent pod identity and lifecycle management that still make it a thing. I do volumeless statefulsets all the time.

> Imagine following usecase. I do want to run with an external database, that means I want to disable the persistence for sqlite, but I still want to run persistence for the keys and tornjak. How would I achieve that with this change?

The same way as today. You just don't turn off persistence. Nothing related to this has changed in this pr. Just the name of the key used to do the enabling/disabling as been renamed to better reflect the way its always worked, and to follow helm convention.

Comment (kfox1111) [reply charts/spire/charts/spire-server/README.md:88]: Thought of one other thing, and I think its really important....

We can't add more volumes to the statefulset without a lot of upgrade pain on the users. Apparently volumeClaimTemplates is immutable. So if the user wanted to start out without, say, Tornjak, and then enable it later they would have to destroy their statefulset if it had sepaerate pvc's. 
```
The StatefulSet "test" is invalid: spec: Forbidden: updates to statefulset spec for fields other than
'replicas','template', 'updateStrategy', 'persistentVolumeClaimRetentionPolicy' and
'minReadySeconds' are forbidden
```

Comment (faisal-memon) [reply charts/spire/charts/spire-server/README.md:88]: We can still use a single pvc as long as they all store persistent data to `/run/spire/data` correct?  Thats what we are doing currently for sqllite3 and the disk keymanager. 

Im not familiar with Tornjak. If Tornjak is in same Pod then it would just need to do the same.

> Oh and if we are about to disable persistence, then probably the statefulset should also be changed to a deployment if persistence is disabled. That is something I have seen in other charts as well. Happy to create a PR to patch those changes on top of this PR.

I think this is a good idea to merge after this PR.

Comment (marcofranssen) [reply charts/spire/charts/spire-server/README.md:88]: > Thought of one other thing, and I think its really important....
> 
> We can't add more volumes to the statefulset without a lot of upgrade pain on the users. Apparently volumeClaimTemplates is immutable. So if the user wanted to start out without, say, Tornjak, and then enable it later they would have to destroy their statefulset if it had sepaerate pvc's.
> 
> ```
> The StatefulSet "test" is invalid: spec: Forbidden: updates to statefulset spec for fields other than
> 'replicas','template', 'updateStrategy', 'persistentVolumeClaimRetentionPolicy' and
> 'minReadySeconds' are forbidden
> ```

Fair point. Maybe we should switch to using a deployment with a separate PVC like they do in this chart. Probably for the reason you mentioning here? 
https://github.com/sigstore/helm-charts/blob/main/charts/rekor/templates/server/pvc.yaml

What do you think? Just trying to prevent API changes in followup releases.

Comment (faisal-memon) [reply charts/spire/charts/spire-server/README.md:88]: I think the discussion on StatefulSet vs Deployment is out of scope for this pr. I think its worth investigating separately as a follow on. For the changes in the pr, are we set on a single PV? If so then i think the changes here are good to go.

**Review by marcofranssen — APPROVED** [review]
Since there is concensus in the approach by other maintainers I'll go ahead merging this one. 🤞 we won't have to change the values structure for persistence later to not introduce another breaking change on the public API.

**Pull request merged**
